### PR TITLE
ENG-1351 SQL Conditions

### PR DESCRIPTION
**Pull request by JadeCara**
Closes [1351](https://ethyca.atlassian.net/browse/ENG-1351?atlOrigin=eyJpIjoiMjZhYzdlMmY1YTNlNDI5YjlkMTYxNzNmNTQxNTE1YWYiLCJwIjoiaiJ9)

### Description Of Changes

This PR implements a SQL parsing and query generation system for conditional dependencies in the Fides backend. The new SQLConditionTranslator enables dynamic evaluation of simple and complex conditional logic against database entities using SQLAlchemy ORM.

#### Core SQL Translation Engine

- SQLConditionTranslator: Main class that converts conditional dependency trees into executable SQLAlchemy queries
- Relationship Traversal: Intelligent JOIN generation using SQLAlchemy relationships with breadth-first search for multi-hop connections
- Field Address Parsing: Robust parsing of dot-notation field addresses (table.column.json_path) with JSON support
- Operator Mapping: Complete operator support including equality, comparison, existence, and list operations

#### Advanced Query Capabilities

- Multi-table Queries: Automatic JOIN discovery and optimization across related tables
- JSON Path Support: Native PostgreSQL JSON operators for nested data access
- Relationship Conditions: Support for conditions on related entities using SQLAlchemy's any() and has() methods
- Reverse Relationship Handling: Intelligent detection and handling of bidirectional relationships

#### Safety & Performance
- SQL Injection Protection: Parameterized queries and safe column reference handling
- Query Optimization: Leverages SQLAlchemy's built-in JOIN optimization and query caching
- Error Handling: Comprehensive error messages with suggestions for common issues
- Type Safety: Full type hints and runtime validation

#### Query Generation Flow
- Condition Parsing: Extract field addresses from condition trees
- Model Discovery: Map table names to SQLAlchemy model classes
- Relationship Analysis: Find JOIN paths using BFS traversal
- Query Building: Generate base query with appropriate JOINs
- Filter Application: Convert conditions to SQLAlchemy filter expressions

#### Key Algorithms
- Breadth-First Search: For finding shortest relationship paths between models
- Relationship Path Caching: Efficient model-to-model mapping with 15-hop depth limit
- Primary Table Detection: Deterministic selection based on condition order

#### Manual Task Conditional Dependencies
- The primary use case is controlling manual task execution based on dynamic conditions:
- User Permission Checks: Execute tasks only for users with specific roles
- Request Status Filtering: Process tasks based on privacy request states
- Data Validation: Skip tasks when certain data conditions aren't met
- Business Logic: Complex multi-table conditions for workflow control

#### Integration Points
- Manual Task Graph: Evaluates conditions during task execution
- Privacy Request Processing: Controls task flow based on request data
- User Management: Role-based task assignment and execution
- Connection Configuration: Task availability based on system state

#### 🧪 Testing Strategy
- Unit Tests (1,242 lines)
  - Mock-based testing for all core functionality 
  - Edge case coverage for relationship traversal
  - Error condition validation
  - Operator mapping verification
- Integration Tests (370 lines)
  - Multi-table query validation
  - Performance testing with complex relationships
  - End-to-end workflow testing
- Demo Scenarios (1,369 lines)
  - Complete user management setup
  - Manual task configuration examples
  - Real-world conditional dependency patterns
  - Interactive testing and validation

### 🔒 Security Considerations
- Parameterized Queries: All user input is properly parameterized
- Column Validation: Strict validation of field addresses against model schemas
- JSON Path Sanitization: Safe handling of JSON path components
- Access Control: Integration with existing Fides permission system

### 📈 Performance Characteristics
- Query Optimization: Leverages SQLAlchemy's JOIN optimization
- Relationship Caching: Model mapping cached for performance
- Depth Limiting: BFS traversal limited to 15 hops to prevent infinite loops
- Lazy Loading: Relationships loaded only when needed

#### The architecture supports easy extension for:
Additional operators and data types
Custom relationship traversal strategies
Database-specific optimizations
Async query execution
Query result caching

### Code Changes

* New Files
  * src/fides/api/task/conditional_dependencies/sql_translator.py (769 lines) - Main translator implementation
  * src/fides/api/task/conditional_dependencies/sql_schemas.py (204 lines) - Field address parsing and operator mapping
  * tests/api/task/conditional_dependencies/test_sql_translator.py (1,242 lines) - Unit tests with mocking 1
  * tests/api/task/conditional_dependencies/test_sql_translator_integration.py (370 lines) - Integration tests against real database
  * tests/api/task/conditional_dependencies/test_sql_schemas.py (484 lines) - Schema validation tests

### Steps to Confirm

1.  Demo & QA
qa/scenarios/sql_translator_demo.py (1,369 lines) - Comprehensive demo showing real-world usage patterns
Run `python qa/scenarios/sql_translator_demo.py setup`
It will set up Users, Manual Tasks, Privacy Request and then run test conditions against them:
```bash
================================================================================

Step 11: Demonstrating SQL Translator functionality...

================================================================================
🔍 SQL TRANSLATOR DEMONSTRATION
================================================================================

📋 Demo 1: Simple Condition
----------------------------------------
📝 Condition: fidesuser.username eq 'demo_owner_user'
   Description: Find user with username 'demo_owner_user'
2025-09-16 23:34:23.670 | DEBUG    | fides.api.task.conditional_dependencies.sql_translator:_build_leaf_filter:623 - Found column username on FidesUser

🔍 Generated SQL:
   SELECT fidesuser.id, fidesuser.created_at, fidesuser.updated_at, fidesuser.username, fidesuser.email_address, fidesuser.first_name, fidesuser.last_name, fidesuser.hashed_password, fidesuser.salt, fidesuser.disabled, fidesuser.disabled_reason, fidesuser.last_login_at, fidesuser.password_reset_at, fidesuser.password_login_enabled, fidesuser.totp_secret 
FROM fidesuser 
WHERE fidesuser.username = 'demo_owner_user'

📊 Results: Found 1 record(s)
   - User: demo_owner_user (demo.owner@example.com)

✅ Expected: 1 user (demo_owner_user)
✅ Actual: 1 record(s)

📋 Demo 2: Group Condition (AND)
----------------------------------------
📝 Condition: AND group
   - fidesuser.username starts_with 'demo_'
   - fidesuser.email_address contains '@example.com'
   Description: Find demo users with example.com email addresses
2025-09-16 23:34:23.671 | DEBUG    | fides.api.task.conditional_dependencies.sql_translator:_build_leaf_filter:623 - Found column username on FidesUser
2025-09-16 23:34:23.671 | DEBUG    | fides.api.task.conditional_dependencies.sql_translator:_build_leaf_filter:623 - Found column email_address on FidesUser

🔍 Generated SQL:
   SELECT fidesuser.id, fidesuser.created_at, fidesuser.updated_at, fidesuser.username, fidesuser.email_address, fidesuser.first_name, fidesuser.last_name, fidesuser.hashed_password, fidesuser.salt, fidesuser.disabled, fidesuser.disabled_reason, fidesuser.last_login_at, fidesuser.password_reset_at, fidesuser.password_login_enabled, fidesuser.totp_secret 
FROM fidesuser 
WHERE fidesuser.username LIKE 'demo_%' AND fidesuser.email_address LIKE '%@example.com%'

📊 Results: Found 5 record(s)
   - User: demo_owner_user (demo.owner@example.com)
   - User: demo_contributor_user (demo.contributor@example.com)
   - User: demo_viewer_user (demo.viewer@example.com)
   ... and 2 more

✅ Expected: 5 users (all demo users)
✅ Actual: 5 record(s)

📋 Demo 3: Nested Condition (AND with OR nested)
----------------------------------------
📝 Condition: Nested AND/OR
   (username = 'demo_owner_user' OR username = 'demo_viewer_user')
   AND email contains '@example.com'
   Description: Find owner or viewer users with example.com emails
2025-09-16 23:34:23.673 | DEBUG    | fides.api.task.conditional_dependencies.sql_translator:_build_leaf_filter:623 - Found column username on FidesUser
2025-09-16 23:34:23.673 | DEBUG    | fides.api.task.conditional_dependencies.sql_translator:_build_leaf_filter:623 - Found column username on FidesUser
2025-09-16 23:34:23.673 | DEBUG    | fides.api.task.conditional_dependencies.sql_translator:_build_leaf_filter:623 - Found column email_address on FidesUser

🔍 Generated SQL:
   SELECT fidesuser.id, fidesuser.created_at, fidesuser.updated_at, fidesuser.username, fidesuser.email_address, fidesuser.first_name, fidesuser.last_name, fidesuser.hashed_password, fidesuser.salt, fidesuser.disabled, fidesuser.disabled_reason, fidesuser.last_login_at, fidesuser.password_reset_at, fidesuser.password_login_enabled, fidesuser.totp_secret 
FROM fidesuser 
WHERE (fidesuser.username = 'demo_owner_user' OR fidesuser.username = 'demo_viewer_user') AND fidesuser.email_address LIKE '%@example.com%'

📊 Results: Found 2 record(s)
   - User: demo_owner_user (demo.owner@example.com)
   - User: demo_viewer_user (demo.viewer@example.com)

✅ Expected: 2 users (demo_owner_user and demo_viewer_user)
✅ Actual: 2 record(s)

📋 Demo 4: Complex Multi-Table Query (Manual Task Business Logic)
----------------------------------------
📝 Condition: Complex Multi-Table Query
   - manual_task_instance.status != 'complete' (PRIMARY)
   - manual_task.references.reference_id = demo_owner_user_id
   - privacyrequest.status != 'complete'
   Description: Find actionable manual task instances for demo_owner_user


🔍 Generated SQL:
   SELECT manual_task_instance.id, manual_task_instance.created_at, manual_task_instance.updated_at, manual_task_instance.task_id, manual_task_instance.config_id, manual_task_instance.entity_id, manual_task_instance.entity_type, manual_task_instance.status, manual_task_instance.completed_at, manual_task_instance.completed_by_id, manual_task_instance.due_date 
FROM manual_task_instance JOIN manual_task ON manual_task.id = manual_task_instance.task_id JOIN privacyrequest ON manual_task_instance.entity_id = privacyrequest.id AND manual_task_instance.entity_type = 'privacy_request' 
WHERE manual_task_instance.status != 'complete' AND (EXISTS (SELECT 1 
FROM manual_task_reference 
WHERE manual_task.id = manual_task_reference.task_id AND manual_task_reference.reference_id = 'fid_7457dafa-5a74-49e4-a320-560d37afd1ab')) AND privacyrequest.status != 'complete'

📊 Results: Found 1 record(s)
   - ManualTaskInstance (ID: man_60656730-a606-4707-af02-af89170c4d08)
     └── Status: pending

✅ Expected: 1 | Actual: 1
🎉 Perfect match! Complex multi-table query handled correctly

```

### Pre-Merge Checklist

* [ ] Issue requirements met
* [ ] All CI pipelines succeeded
* [ ] `CHANGELOG.md` updated
  * [ ] Add a https://github.com/ethyca/fides/labels/db-migration label to the entry if your change includes a DB migration
  * [ ] Add a https://github.com/ethyca/fides/labels/high-risk label to the entry if your change includes a high-risk change (i.e. potential for performance impact or unexpected regression) that should be flagged
  * [ ] Updates unreleased work already in Changelog, no new entry necessary
* Followup issues:
  * [ ] Followup issues created
  * [ ] No followup issues
* Database migrations:
  * [ ] Ensure that your downrev is up to date with the latest revision on `main`
  * [ ] Ensure that your `downgrade()` migration is correct and works
    * [ ] If a downgrade migration is not possible for this change, please call this out in the PR description!
  * [ ] No migrations
* Documentation:
  * [ ] Documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] Documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
  * [ ] If there are any new client scopes created as part of the pull request, remember to update public-facing documentation that references our scope registry
  * [ ] No documentation updates required

### CODE REVIEW

File: tests/api/task/conditional_dependencies/test_sql_translator.py [hunk 1]
```diff
@@ -0,0 +1,1242 @@
...
+            "FROM fidesuserpermissions ",
+            "WHERE fidesuser.id = fidesuserpermissions.user_id AND 'approver' = ANY (fidesuserpermissions.roles))) OR (EXISTS (SELECT 1 ",
+            "FROM fidesuserpermissions ",
+            "WHERE fidesuser.id = fidesuserpermissions.user_id AND 'approver' = ANY (fidesuserpermissions.roles))) AND (fidesuser.password_login_enabled = true OR fidesuser.totp_secret IS NOT NULL)",
+        ]
+
+        for part in expected_parts:
+            assert (
+                part in query_str
+            ), f"Expected '{part}' not found in generated SQL: {query_str}"
```
[the remark this answers is not in the document]
Comment (JadeCara) [reply tests/api/task/conditional_dependencies/test_sql_translator.py:648]: Going to skip this one - There are other tests that are more focused on specific functions this test is intended to verify these functions work together in an expected way.

File: src/fides/api/task/conditional_dependencies/sql_schemas.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        """
+        is_json_path = self.json_path is not None and len(self.json_path) > 0
+        if not is_json_path or not enable_json_operators:
+            return self.column_name
+
+        if self.json_path is None:
+            # This should never happen
+            raise SQLTranslationError(
+                "Field address internal error."
+            )  # pragma: no cover
```
[the remark this answers is not in the document]
Comment (JadeCara) [reply src/fides/api/task/conditional_dependencies/sql_schemas.py:199]: Agreed, but mypy still yells about it.

File: tests/api/task/conditional_dependencies/test_sql_schemas.py [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+            OPERATOR_MAP[Operator.eq] = (
+                lambda x, y: x == y
+            )  # This should work but not affect the original
+        except TypeError:
+            # If it's immutable, that's good
+            pass
+
+        # Verify the original is unchanged
+        assert set(OPERATOR_MAP.keys()) == original_keys
+        assert len(OPERATOR_MAP) == original_size
```
[the remark this answers is not in the document]
Comment (JadeCara) [reply tests/api/task/conditional_dependencies/test_sql_schemas.py:468]: Updated to use proxy map for actual immutability.

File: src/fides/api/task/conditional_dependencies/sql_translator.py [hunk 4]
```diff
@@ -0,0 +1,748 @@
...
+        Args:
+            condition: The condition to build the filter expression for
+            model_classes: Dictionary of table names to model classes
+
+        Returns:
+            The SQLAlchemy filter expression, or None if no filter expression can be built
+        """
+        field_addr = FieldAddress.parse(condition.field_address)
+
+        if not field_addr.table_name or field_addr.table_name not in model_classes:
```
Comment (thabofletcher) [inline src/fides/api/task/conditional_dependencies/sql_translator.py:575]: Is this or the other "return None" condition worthy of a warning or perhaps erroring/SQLTranslationError

If I understand correctly this would cause badly written input (wrong field address or something) to not contribute to the predicated, but could it cause something to be silently missed? Or an unintentional query with a missed conditional?

**Review by thabofletcher — APPROVED** [review]
@JadeCara overall this looks great! 

My only concern with this is it is big change and by talking to SQL and writing a translation engine there are some new security attack surfaces and potential bugs. It looks like we have sql injection covered (and a bunch of other stuff thought through), which is great 👍 🤞 This is also nicely decoupled so lessens the risk.

I had one question in the code and then also wanted to point out codecov - I'm not sure if we are expected to test every line or not but that would make this even larger

**Comment by JadeCara** [comment]
> My only concern with this is it is big change and by talking to SQL and writing a translation engine there are some new security attack surfaces and potential bugs. It looks like we have sql injection covered (and a bunch of other stuff thought through), which is great 👍 🤞 This is also nicely decoupled so lessens the risk.

I put in some extra validation around: 
- SQL Injection: All user input is properly parameterized and escaped
- Pattern Injection: LIKE wildcards are escaped to prevent unintended matches
- JSON Path Injection: PostgreSQL JSON operators are protected from quote-breaking
- Performance Attacks: Wildcard escaping prevents expensive pattern operations
- Information Disclosure: Pattern probing attacks are blocked
